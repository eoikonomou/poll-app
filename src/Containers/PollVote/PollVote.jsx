import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { observer } from 'mobx-react';
import './PollVote.scss';

const PollVote = observer(({ store }) => {
  const [selectedOption, setSelectedOption] = useState(null);
  const {
    pollQuestion,
    pollOptions,
    addVote,
    pollLocked,
  } = store;

  const castVote = () => {
    addVote(selectedOption);
    setSelectedOption(null);
  };

  const onSelect = (optionId) => () => {
    setSelectedOption(optionId);
  };

  return (
    <div className="section">
      <div className="section-title">{pollQuestion || 'Enter poll question'}</div>
      <div className="flex-container">
        {pollOptions.map((option) => (
          <div className="vote-option-container" key={option.id}>
            <input
              type="radio"
              id={option.id}
              name={option.id}
              value={option.id}
              checked={selectedOption === option.id}
              onChange={onSelect(option.id)}
            />
            <label htmlFor={option.id}>{option.value}</label>
          </div>
        ))}
      </div>
      <div className="footer">
        <button
          type="button"
          disabled={selectedOption === null || !pollLocked}
          className="vote-btn"
          onClick={castVote}
        >
          Vote
        </button>
        {!pollLocked && <div className="footer-info">Lock poll options to enable voting</div>}
      </div>
    </div>
  );
});

PollVote.displayName = 'PollVote';

PollVote.propTypes = {
  store: PropTypes.objectOf(PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.func,
    PropTypes.array,
    PropTypes.bool,
  ])).isRequired,
};

export default PollVote;
