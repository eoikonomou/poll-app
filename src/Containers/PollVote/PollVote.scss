.vote-option-container {
  align-items: center;
  padding: 10px;

  * {
    vertical-align: middle;
    margin-right: 15px;
  }
}

.footer-info {
  padding-left: 20px;
}
