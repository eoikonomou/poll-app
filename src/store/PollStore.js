import { action, observable } from 'mobx';

export default class PollStore {
  @observable pollQuestion = '';

  @observable pollOptions = [];

  @observable pollLocked = false;

  @action setPollQuestion = (data) => {
    this.pollQuestion = data;
  }

  @action setPollOptions = (data) => {
    this.pollOptions = data;
  }

  @action addPollOption = (data) => {
    this.pollOptions.push(data);
  }

  @action addVote = (optionId) => {
    const option = this.pollOptions.find((opt) => opt.id === optionId);
    option.count += 1;
  }

  @action reset = () => {
    this.pollOptions = [];
    this.pollQuestion = '';
    this.pollLocked = false;
  }

  @action setPollLocked = (data) => {
    this.pollLocked = data;
  }
}
