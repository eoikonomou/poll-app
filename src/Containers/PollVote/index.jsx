import PollVote from './PollVote';

export default PollVote;
