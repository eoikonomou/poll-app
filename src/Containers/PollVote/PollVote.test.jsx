import React from 'react';
import { shallow, mount } from 'enzyme';
import PollVote from './PollVote';
import PollStore from '../../store/PollStore';

describe('PollVote', () => {
  let store = null;
  let wrapper = null;

  beforeEach(() => {
    store = new PollStore();
  });

  afterEach(() => {
    store = null;
    wrapper = null;
  });

  test('should render successfully', () => {
    wrapper = shallow(<PollVote store={store} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('should increase the number of input fields when poll options increase', () => {
    wrapper = shallow(<PollVote store={store} />);
    expect(wrapper.find('input')).toHaveLength(0);
    store.setPollOptions([{ id: 'opt1', value: 'opt1', count: 0 }]);
    wrapper = shallow(<PollVote store={store} />);
    expect(wrapper.find('input')).toHaveLength(1);
    store.setPollOptions([
      { id: 'opt1', value: 'opt1', count: 0 },
      { id: 'opt2', value: 'opt2', count: 0 },
      { id: 'opt3', value: 'opt3', count: 0 },
      { id: 'opt4', value: 'opt4', count: 0 },
    ]);
    wrapper = shallow(<PollVote store={store} />);
    expect(wrapper.find('input')).toHaveLength(4);
  });

  test('should have .vote-btn button disabled while there is no selected option or pollLocked is false', () => {
    wrapper = shallow(<PollVote store={store} />);
    expect(wrapper.find('button').prop('disabled')).toBe(true);
    store.setPollLocked(true);
    wrapper = shallow(<PollVote store={store} />);
    expect(wrapper.find('button').prop('disabled')).toBe(true);
    store.setPollOptions([
      { id: 'opt1', value: 'opt1', count: 0 },
      { id: 'opt2', value: 'opt2', count: 0 },
      { id: 'opt3', value: 'opt3', count: 0 },
      { id: 'opt4', value: 'opt4', count: 0 },
    ]);
    wrapper = mount(<PollVote store={store} />);
    const radioButton = () => wrapper.find('input').at(0);
    radioButton().simulate('change', { target: { checked: true } });
    expect(wrapper.find('button').prop('disabled')).toBe(false);
  });

  test('should increase the count property of the option by pressing the vote button with a selected option', () => {
    const spy = jest.spyOn(store, 'addVote');
    store.setPollLocked(true);
    store.setPollOptions([
      { id: 'opt1', value: 'opt1', count: 0 },
      { id: 'opt2', value: 'opt2', count: 0 },
      { id: 'opt3', value: 'opt3', count: 0 },
      { id: 'opt4', value: 'opt4', count: 0 },
    ]);
    wrapper = mount(<PollVote store={store} />);
    const radioButton = () => wrapper.find('input').at(0);
    radioButton().simulate('change', { target: { checked: true } });
    const voteButton = () => wrapper.find('button').at(0);
    voteButton().simulate('click');
    expect(spy).toHaveBeenCalledTimes(1);
    expect(store.pollOptions[0].count).toEqual(1);
  });

  test('should unselect the selected option after voting', () => {
    store.setPollLocked(true);
    store.setPollOptions([
      { id: 'opt1', value: 'opt1', count: 0 },
      { id: 'opt2', value: 'opt2', count: 0 },
      { id: 'opt3', value: 'opt3', count: 0 },
      { id: 'opt4', value: 'opt4', count: 0 },
    ]);
    wrapper = mount(<PollVote store={store} />);
    const radioButton = () => wrapper.find('input').at(0);
    radioButton().simulate('change', { target: { checked: true } });
    const voteButton = () => wrapper.find('button').at(0);
    voteButton().simulate('click');
    expect(radioButton().prop('checked')).toBe(false);
  });

  test('should change the section-title when pollQuestion changes', () => {
    wrapper = shallow(<PollVote store={store} />);
    const title = () => wrapper.find('.section-title').at(0);
    const firstTitle = title().text();
    expect(firstTitle).toEqual('Enter poll question');
    store.setPollQuestion('test');
    wrapper = shallow(<PollVote store={store} />);
    const secondTitle = title().text();
    expect(secondTitle).toEqual('test');
  });
});
