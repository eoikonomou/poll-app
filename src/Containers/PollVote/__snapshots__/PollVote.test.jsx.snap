// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PollVote should render successfully 1`] = `
<div
  className="section"
>
  <div
    className="section-title"
  >
    Enter poll question
  </div>
  <div
    className="flex-container"
  />
  <div
    className="footer"
  >
    <button
      className="vote-btn"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      Vote
    </button>
    <div
      className="footer-info"
    >
      Lock poll options to enable voting
    </div>
  </div>
</div>
`;
